File: frontend/src/api/alerts.ts
/**
 * Alerts API service
 * Provides functions for managing alert rules, destinations, routes, and delivery history
 */

import { getErrorMessage } from '@/utils/errorHandling';

// Create API client instance
const API_BASE_URL =
  import.meta.env?.VITE_API_URL || 'http://localhost:8080/api/v1';

// Types for alerts system
export interface AlertRule {
  id: number;
  name: string;
  description?: string;
  query: string;
  condition:
    | 'greater'
    | 'greater_equal'
    | 'less'
    | 'less_equal'
    | 'equal'
    | 'not_equal';
  threshold: number;
  interval: string; // duration string like "5m"
  for?: string; // duration string like "10m"
  labels?: Record<string, string>;
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface AlertDestination {
  id: number;
  name: string;
  description?: string;
  type: 'webhook' | 'slack' | 'pushover';
  config: Record<string, any>;
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface AlertRoute {
  id: number;
  name: string;
  description?: string;
  destination_id: number;
  matchers: Record<string, string>;
  priority: number;
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface Alert {
  id: number;
  rule_id: number;
  entity_id: string;
  entity_type: string;
  dedupe_key: string;
  status: 'firing' | 'resolved';
  value?: number;
  starts_at: string;
  ends_at?: string;
  labels?: Record<string, string>;
  annotations?: Record<string, string>;
  rule?: AlertRule;
}

export interface AlertDelivery {
  id: number;
  alert_id: number;
  destination_id: number;
  route_id: number;
  status: 'pending' | 'delivered' | 'failed';
  attempt_count: number;
  max_attempts: number;
  next_attempt_at?: string;
  error_message?: string;
  request_payload?: string;
  response_payload?: string;
  response_status?: number;
  created_at: string;
  updated_at: string;
}

export interface EngineStats {
  enabled: boolean;
  evaluation_interval: string;
  queue_size: number;
  rules: number;
  destinations: number;
  alerts: {
    total_alerts: number;
    firing_alerts: number;
    resolved_alerts: number;
    active_rules: number;
    affected_entities: number;
  };
  deliveries: {
    total_deliveries: number;
    successful_deliveries: number;
    failed_deliveries: number;
    pending_deliveries: number;
    avg_attempts: number;
  };
  service: {
    total_processed: number;
    successful_sent: number;
    failed: number;
    retries: number;
    currently_pending: number;
    average_latency: number;
    last_processed_at: string;
    workers_running: number;
  };
  providers_enabled: string[];
}

// API Request/Response types
export interface CreateAlertRuleParams {
  name: string;
  description?: string;
  query: string;
  condition: string;
  threshold: number;
  interval: string;
  for?: string;
  labels?: Record<string, string>;
  is_enabled?: boolean;
}

export interface UpdateAlertRuleParams {
  name?: string;
  description?: string;
  query?: string;
  condition?: string;
  threshold?: number;
  interval?: string;
  for?: string;
  labels?: Record<string, string>;
  is_enabled?: boolean;
}

export interface CreateAlertDestinationParams {
  name: string;
  description?: string;
  type: string;
  config: Record<string, any>;
  is_enabled?: boolean;
}

export interface UpdateAlertDestinationParams {
  name?: string;
  description?: string;
  type?: string;
  config?: Record<string, any>;
  is_enabled?: boolean;
}

export interface CreateAlertRouteParams {
  name: string;
  description?: string;
  destination_id: number;
  matchers: Record<string, string>;
  priority?: number;
  is_enabled?: boolean;
}

export interface UpdateAlertRouteParams {
  name?: string;
  description?: string;
  destination_id?: number;
  matchers?: Record<string, string>;
  priority?: number;
  is_enabled?: boolean;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    limit: number;
    offset: number;
    total: number;
  };
}

export interface TestDestinationRequest {
  message: string;
}

// Helper function for API calls
async function apiCall<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const url = `${API_BASE_URL}${endpoint}`;
  const config: RequestInit = {
    headers: {
      'Content-Type': 'application/json',
      ...options.headers,
    },
    ...options,
  };

  try {
    const response = await fetch(url, config);

    if (!response.ok) {
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      try {
        const errorData = await response.json();
        if (errorData.error) {
          errorMessage = errorData.error;
        }
      } catch {
        // Failed to parse error response, use default message
      }
      throw new Error(errorMessage);
    }

    // Handle 204 No Content responses
    if (response.status === 204) {
      return undefined as T;
    }

    return await response.json();
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

// Alert Rules API
export const alertRulesApi = {
  async list(
    params: { limit?: number; offset?: number; enabled?: boolean } = {},
  ): Promise<PaginatedResponse<AlertRule>> {
    const searchParams = new URLSearchParams();
    if (params.limit) searchParams.set('limit', params.limit.toString());
    if (params.offset) searchParams.set('offset', params.offset.toString());
    if (params.enabled !== undefined)
      searchParams.set('enabled', params.enabled.toString());

    const query = searchParams.toString();
    return apiCall(`/alerts/rules${query ? `?${query}` : ''}`);
  },

  async get(id: number): Promise<AlertRule> {
    return apiCall(`/alerts/rules/${id}`);
  },

  async create(params: CreateAlertRuleParams): Promise<AlertRule> {
    return apiCall(`/alerts/rules`, {
      method: 'POST',
      body: JSON.stringify(params),
    });
  },

  async update(id: number, params: UpdateAlertRuleParams): Promise<void> {
    return apiCall(`/alerts/rules/${id}`, {
      method: 'PUT',
      body: JSON.stringify(params),
    });
  },

  async delete(id: number): Promise<void> {
    return apiCall(`/alerts/rules/${id}`, {
      method: 'DELETE',
    });
  },

  async test(
    id: number,
  ): Promise<{ rule: AlertRule; results: any; tested_at: string }> {
    return apiCall(`/alerts/rules/${id}/test`, {
      method: 'POST',
    });
  },
};

// Alert Destinations API
export const alertDestinationsApi = {
  async list(
    params: { limit?: number; offset?: number; enabled?: boolean } = {},
  ): Promise<PaginatedResponse<AlertDestination>> {
    const searchParams = new URLSearchParams();
    if (params.limit) searchParams.set('limit', params.limit.toString());
    if (params.offset) searchParams.set('offset', params.offset.toString());
    if (params.enabled !== undefined)
      searchParams.set('enabled', params.enabled.toString());

    const query = searchParams.toString();
    return apiCall(`/alerts/destinations${query ? `?${query}` : ''}`);
  },

  async get(id: number): Promise<AlertDestination> {
    return apiCall(`/alerts/destinations/${id}`);
  },

  async create(
    params: CreateAlertDestinationParams,
  ): Promise<AlertDestination> {
    return apiCall(`/alerts/destinations`, {
      method: 'POST',
      body: JSON.stringify(params),
    });
  },

  async update(
    id: number,
    params: UpdateAlertDestinationParams,
  ): Promise<void> {
    return apiCall(`/alerts/destinations/${id}`, {
      method: 'PUT',
      body: JSON.stringify(params),
    });
  },

  async delete(id: number): Promise<void> {
    return apiCall(`/alerts/destinations/${id}`, {
      method: 'DELETE',
    });
  },

  async test(
    id: number,
    params: TestDestinationRequest,
  ): Promise<{ message: string; tested_at: string }> {
    return apiCall(`/alerts/destinations/${id}/test`, {
      method: 'POST',
      body: JSON.stringify(params),
    });
  },
};

// Alert Routes API
export const alertRoutesApi = {
  async list(
    params: { limit?: number; offset?: number; destination_id?: number } = {},
  ): Promise<PaginatedResponse<AlertRoute>> {
    const searchParams = new URLSearchParams();
    if (params.limit) searchParams.set('limit', params.limit.toString());
    if (params.offset) searchParams.set('offset', params.offset.toString());
    if (params.destination_id)
      searchParams.set('destination_id', params.destination_id.toString());

    const query = searchParams.toString();
    return apiCall(`/alerts/routes${query ? `?${query}` : ''}`);
  },

  async get(id: number): Promise<AlertRoute> {
    return apiCall(`/alerts/routes/${id}`);
  },

  async create(params: CreateAlertRouteParams): Promise<AlertRoute> {
    return apiCall(`/alerts/routes`, {
      method: 'POST',
      body: JSON.stringify(params),
    });
  },

  async update(id: number, params: UpdateAlertRouteParams): Promise<void> {
    return apiCall(`/alerts/routes/${id}`, {
      method: 'PUT',
      body: JSON.stringify(params),
    });
  },

  async delete(id: number): Promise<void> {
    return apiCall(`/alerts/routes/${id}`, {
      method: 'DELETE',
    });
  },
};

// Alerts API
export const alertsApi = {
  async list(
    params: {
      limit?: number;
      offset?: number;
      status?: string;
      rule_id?: number;
    } = {},
  ): Promise<PaginatedResponse<Alert>> {
    const searchParams = new URLSearchParams();
    if (params.limit) searchParams.set('limit', params.limit.toString());
    if (params.offset) searchParams.set('offset', params.offset.toString());
    if (params.status) searchParams.set('status', params.status);
    if (params.rule_id) searchParams.set('rule_id', params.rule_id.toString());

    const query = searchParams.toString();
    return apiCall(`/alerts${query ? `?${query}` : ''}`);
  },

  async get(id: number): Promise<Alert> {
    return apiCall(`/alerts/${id}`);
  },
};

// Alert Deliveries API
export const alertDeliveriesApi = {
  async list(
    params: {
      limit?: number;
      offset?: number;
      alert_id?: number;
      destination_id?: number;
      status?: string;
    } = {},
  ): Promise<PaginatedResponse<AlertDelivery>> {
    const searchParams = new URLSearchParams();
    if (params.limit) searchParams.set('limit', params.limit.toString());
    if (params.offset) searchParams.set('offset', params.offset.toString());
    if (params.alert_id)
      searchParams.set('alert_id', params.alert_id.toString());
    if (params.destination_id)
      searchParams.set('destination_id', params.destination_id.toString());
    if (params.status) searchParams.set('status', params.status);

    const query = searchParams.toString();
    return apiCall(`/alerts/deliveries${query ? `?${query}` : ''}`);
  },
};

// Engine API
export const alertEngineApi = {
  async getStatus(): Promise<{ engine: EngineStats; retrieved_at: string }> {
    return apiCall(`/alerts/engine/status`);
  },

  async triggerEvaluation(): Promise<{
    message: string;
    triggered_at: string;
  }> {
    return apiCall(`/alerts/engine/evaluate`, {
      method: 'POST',
    });
  },
};

File: internal/models/alerts.go
// Package models contains domain models for the alerts system
package models

import (
	"encoding/json"
	"time"
)

// ==============================================================================
// ALERT SYSTEM CONSTANTS
// ==============================================================================

// Alert condition types
const (
	AlertConditionGreater      = "gt"
	AlertConditionGreaterEqual = "gte"
	AlertConditionLess         = "lt"
	AlertConditionLessEqual    = "lte"
	AlertConditionEqual        = "eq"
	AlertConditionNotEqual     = "ne"
)

// ==============================================================================
// ALERT SYSTEM DOMAIN MODELS
// ==============================================================================

// AlertRule represents a rule that triggers alerts based on conditions
type AlertRule struct {
	ID          int64             `json:"id"`
	Name        string            `json:"name"`
	Description *string           `json:"description,omitempty"`
	Query       string            `json:"query"`
	Condition   string            `json:"condition"` // e.g., "gt", "lt", "eq"
	Threshold   float64           `json:"threshold"`
	Interval    time.Duration     `json:"interval" swaggertype:"integer" format:"int64" example:"300000000000"`    // How often to evaluate (nanoseconds)
	For         *time.Duration    `json:"for,omitempty" swaggertype:"integer" format:"int64" example:"600000000000"` // How long condition must persist (nanoseconds)
	Labels      map[string]string `json:"labels,omitempty"`
	IsEnabled   bool              `json:"is_enabled"`
	CreatedAt   time.Time         `json:"created_at"`
	UpdatedAt   time.Time         `json:"updated_at"`
}

// Alert represents a triggered alert instance
type Alert struct {
	ID          int64             `json:"id"`
	RuleID      int64             `json:"rule_id"`
	EntityID    string            `json:"entity_id"` // e.g., volume_id, container_id
	EntityType  string            `json:"entity_type"` // e.g., "volume", "container"
	DedupeKey   string            `json:"dedupe_key"` // For deduplication
	Status      string            `json:"status"` // "firing", "resolved"
	Value       *float64          `json:"value,omitempty"`
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
	StartsAt    time.Time         `json:"starts_at"`
	EndsAt      *time.Time        `json:"ends_at,omitempty"`
	CreatedAt   time.Time         `json:"created_at"`
	UpdatedAt   time.Time         `json:"updated_at"`
	
	// Relationships
	Rule *AlertRule `json:"rule,omitempty"`
}

// AlertDestination represents a notification destination
type AlertDestination struct {
	ID         int64                  `json:"id"`
	Name       string                 `json:"name"`
	Type       string                 `json:"type"` // "webhook", "slack", "pushover"
	Config     map[string]interface{} `json:"config"`
	IsEnabled  bool                   `json:"is_enabled"`
	CreatedAt  time.Time              `json:"created_at"`
	UpdatedAt  time.Time              `json:"updated_at"`
}

// AlertRoute represents routing rules for alerts to destinations
type AlertRoute struct {
	ID            int64             `json:"id"`
	Name          string            `json:"name"`
	Matchers      map[string]string `json:"matchers"` // Label matchers
	DestinationID int64             `json:"destination_id"`
	Priority      int32             `json:"priority"` // Lower number = higher priority
	IsEnabled     bool              `json:"is_enabled"`
	CreatedAt     time.Time         `json:"created_at"`
	UpdatedAt     time.Time         `json:"updated_at"`
	
	// Relationships
	Destination *AlertDestination `json:"destination,omitempty"`
}

// AlertDelivery represents a delivery attempt to a destination
type AlertDelivery struct {
	ID               int64      `json:"id"`
	AlertID          int64      `json:"alert_id"`
	DestinationID    int64      `json:"destination_id"`
	RouteID          int64      `json:"route_id"`
	Status           string     `json:"status"` // "pending", "delivered", "failed", "retrying"
	AttemptCount     int32      `json:"attempt_count"`
	MaxAttempts      int32      `json:"max_attempts"`
	NextAttemptAt    *time.Time `json:"next_attempt_at,omitempty"`
	LastAttemptAt    *time.Time `json:"last_attempt_at,omitempty"`
	ErrorMessage     *string    `json:"error_message,omitempty"`
	DeliveredAt      *time.Time `json:"delivered_at,omitempty"`
	RequestPayload   *string    `json:"request_payload,omitempty"`
	ResponsePayload  *string    `json:"response_payload,omitempty"`
	ResponseStatus   *int32     `json:"response_status,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
	
	// Relationships
	Alert       *Alert            `json:"alert,omitempty"`
	Destination *AlertDestination `json:"destination,omitempty"`
	Route       *AlertRoute       `json:"route,omitempty"`
}

// ==============================================================================
// CREATE/UPDATE PARAMETERS
// ==============================================================================

// CreateAlertRuleParams represents parameters for creating an alert rule
type CreateAlertRuleParams struct {
	Name        string            `json:"name" validate:"required,min=1,max=100"`
	Description *string           `json:"description,omitempty" validate:"omitempty,max=500"`
	Query       string            `json:"query" validate:"required"`
	Condition   string            `json:"condition" validate:"required,oneof=gt lt eq ne gte lte"`
	Threshold   float64           `json:"threshold" validate:"required"`
	Interval    time.Duration     `json:"interval" validate:"required,min=1m" swaggertype:"integer" format:"int64" example:"300000000000"`
	For         *time.Duration    `json:"for,omitempty" validate:"omitempty,min=0" swaggertype:"integer" format:"int64" example:"600000000000"`
	Labels      map[string]string `json:"labels,omitempty"`
	IsEnabled   bool              `json:"is_enabled"`
}

// UpdateAlertRuleParams represents parameters for updating an alert rule
type UpdateAlertRuleParams struct {
	ID          int64             `json:"id"`
	Name        string            `json:"name" validate:"required,min=1,max=100"`
	Description *string           `json:"description,omitempty" validate:"omitempty,max=500"`
	Query       string            `json:"query" validate:"required"`
	Condition   string            `json:"condition" validate:"required,oneof=gt lt eq ne gte lte"`
	Threshold   float64           `json:"threshold" validate:"required"`
	Interval    time.Duration     `json:"interval" validate:"required,min=1m" swaggertype:"integer" format:"int64" example:"300000000000"`
	For         *time.Duration    `json:"for,omitempty" validate:"omitempty,min=0" swaggertype:"integer" format:"int64" example:"600000000000"`
	Labels      map[string]string `json:"labels,omitempty"`
	IsEnabled   bool              `json:"is_enabled"`
}

// CreateAlertDestinationParams represents parameters for creating a destination
type CreateAlertDestinationParams struct {
	Name      string                 `json:"name" validate:"required,min=1,max=100"`
	Type      string                 `json:"type" validate:"required,oneof=webhook slack pushover"`
	Config    map[string]interface{} `json:"config" validate:"required"`
	IsEnabled bool                   `json:"is_enabled"`
}

// UpdateAlertDestinationParams represents parameters for updating a destination
type UpdateAlertDestinationParams struct {
	ID        int64                  `json:"id"`
	Name      string                 `json:"name" validate:"required,min=1,max=100"`
	Type      string                 `json:"type" validate:"required,oneof=webhook slack pushover"`
	Config    map[string]interface{} `json:"config" validate:"required"`
	IsEnabled bool                   `json:"is_enabled"`
}

// CreateAlertRouteParams represents parameters for creating a route
type CreateAlertRouteParams struct {
	Name          string            `json:"name" validate:"required,min=1,max=100"`
	Matchers      map[string]string `json:"matchers" validate:"required"`
	DestinationID int64             `json:"destination_id" validate:"required"`
	Priority      int32             `json:"priority" validate:"min=0"`
	IsEnabled     bool              `json:"is_enabled"`
}

// UpdateAlertRouteParams represents parameters for updating a route
type UpdateAlertRouteParams struct {
	ID            int64             `json:"id"`
	Name          string            `json:"name" validate:"required,min=1,max=100"`
	Matchers      map[string]string `json:"matchers" validate:"required"`
	DestinationID int64             `json:"destination_id" validate:"required"`
	Priority      int32             `json:"priority" validate:"min=0"`
	IsEnabled     bool              `json:"is_enabled"`
}

// ==============================================================================
// HELPER TYPES AND CONSTANTS
// ==============================================================================

// Alert status constants
const (
	AlertStatusFiring   = "firing"
	AlertStatusResolved = "resolved"
)

// Alert delivery status constants
const (
	DeliveryStatusPending   = "pending"
	DeliveryStatusDelivered = "delivered"
	DeliveryStatusFailed    = "failed" 
	DeliveryStatusRetrying  = "retrying"
)

// Alert condition constants
const (
	ConditionGreaterThan      = "gt"
	ConditionLessThan         = "lt"
	ConditionEqual            = "eq"
	ConditionNotEqual         = "ne"
	ConditionGreaterThanEqual = "gte"
	ConditionLessThanEqual    = "lte"
)

// Provider type constants
const (
	ProviderTypeWebhook  = "webhook"
	ProviderTypeSlack    = "slack"
	ProviderTypePushover = "pushover"
)

// TestDeliveryRequest represents a request to test a destination
type TestDeliveryRequest struct {
	DestinationID int64  `json:"destination_id" validate:"required"`
	Message       string `json:"message" validate:"required,max=1000"`
}

// TestDeliveryResponse represents the response from testing a destination
type TestDeliveryResponse struct {
	Success        bool    `json:"success"`
	Message        string  `json:"message"`
	ResponseStatus *int32  `json:"response_status,omitempty"`
	ResponseBody   *string `json:"response_body,omitempty"`
	ErrorMessage   *string `json:"error_message,omitempty"`
}

// AlertContext represents the template context for alerts
type AlertContext struct {
	Alert       *Alert            `json:"alert"`
	Rule        *AlertRule        `json:"rule"`
	Value       *float64          `json:"value,omitempty"`
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
}

// Helper methods for JSON marshaling/unmarshaling
func (ar *AlertRule) MarshalLabels() ([]byte, error) {
	return json.Marshal(ar.Labels)
}

func (ar *AlertRule) UnmarshalLabels(data []byte) error {
	return json.Unmarshal(data, &ar.Labels)
}

func (a *Alert) MarshalLabels() ([]byte, error) {
	return json.Marshal(a.Labels)
}

func (a *Alert) UnmarshalLabels(data []byte) error {
	return json.Unmarshal(data, &a.Labels)
}

func (a *Alert) MarshalAnnotations() ([]byte, error) {
	return json.Marshal(a.Annotations)
}

func (a *Alert) UnmarshalAnnotations(data []byte) error {
	return json.Unmarshal(data, &a.Annotations)
}

func (ad *AlertDestination) MarshalConfig() ([]byte, error) {
	return json.Marshal(ad.Config)
}

func (ad *AlertDestination) UnmarshalConfig(data []byte) error {
	return json.Unmarshal(data, &ad.Config)
}

func (ar *AlertRoute) MarshalMatchers() ([]byte, error) {
	return json.Marshal(ar.Matchers)
}

func (ar *AlertRoute) UnmarshalMatchers(data []byte) error {
	return json.Unmarshal(data, &ar.Matchers)
}